examples/developer-tools/src/blocks/factory_base.ts

```diff
@@ -0,0 +1,117 @@
+import * as Blockly from 'blockly';
+
+export const factoryBase = {
+  // Base of new block.
+  init: function() {
+    this.setColour(120);
+    this.appendDummyInput()
+        .appendField('name')
+        .appendField(new Blockly.FieldTextInput('block_type'), 'NAME');
+    this.appendStatementInput('INPUTS')
+        .setCheck('Input')
+        .appendField('inputs');
+    const inputsDropdown = new Blockly.FieldDropdown([
+      ['automatic inputs', 'AUTO'],
+      ['external inputs', 'EXT'],
+      ['inline inputs', 'INT']]);
+    this.appendDummyInput()
+        .appendField(inputsDropdown, 'INLINE');
+    const connectionsDropdown = new Blockly.FieldDropdown([
+      ['no connections', 'NONE'],
+      ['← left output', 'LEFT'],
+      ['↕ top+bottom connections', 'BOTH'],
+      ['↑ top connection', 'TOP'],
+      ['↓ bottom connection', 'BOTTOM']],
+    function(option: string): string {
+      this.getSourceBlock().updateShape(option);
+      // Connect a shadow block to this new input.
+      this.getSourceBlock().spawnOutputShadow(option);
+      return option;
+    });
+    this.appendDummyInput()
+        .appendField(connectionsDropdown, 'CONNECTIONS');
+    this.appendValueInput('TOOLTIP')
+        .setCheck('String')
+        .appendField('tooltip');
+    this.appendValueInput('HELPURL')
+        .setCheck('String')
+        .appendField('help url');
+    this.appendValueInput('COLOUR')
+        .setCheck('Colour')
+        .appendField('colour');
+    this.setTooltip('Build a custom block by plugging\n' +
+        'fields, inputs and other blocks here.');
+    this.setHelpUrl(
+        'https://developers.google.com/blockly/guides/create-custom-blocks/block-factory');
+  },
+  mutationToDom: function() {
+    const container = Blockly.utils.xml.createElement('mutation');
+    container.setAttribute('connections', this.getFieldValue('CONNECTIONS'));
+    return container;
+  },
+  domToMutation: function(xmlElement: Element) {
+    const connections = xmlElement.getAttribute('connections');
+    this.updateShape(connections);
+  },
```

Comment:
Why is this using XML hooks instead of JSON hooks?

Reply:
I started out by copying the block definitions out of the existing project and then updating them. This was removed in a later commit because we don't need a mutation to save the dropdown value at all.